### PR TITLE
75289 create saved filter endpoint

### PR DESCRIPTION
**Pull request by llykkelarsen**
endpoint to create new saved filter

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Command/PersonCommands/CreateSavedFilter/CreateSavedFilterCommandHandler.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using Equinor.Procosys.Preservation.Command.PersonCommands.CreateSavedFilter;
+using Equinor.Procosys.Preservation.Domain;
+using Equinor.Procosys.Preservation.Domain.AggregateModels.PersonAggregate;
+using MediatR;
+using ServiceResult;
+
+namespace Equinor.Procosys.Preservation.Command.ActionCommands.CreateAction
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/PersonCommands/CreateSavedFilter/CreateSavedFilterCommandHandler.cs:9]: wrong ns

File: src/Equinor.Procosys.Preservation.Command/PersonCommands/CreateSavedFilter/CreateSavedFilterCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using Equinor.Procosys.Preservation.Command.PersonCommands.CreateSavedFilter;
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/PersonCommands/CreateSavedFilter/CreateSavedFilterCommandHandler.cs:3]: remove, should be not in use when fixing ns

File: src/Equinor.Procosys.Preservation.Command/Validators/SavedFilterValidators/ISavedFilterValidator.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+
+namespace Equinor.Procosys.Preservation.Command.Validators.SavedFilterValidators
+{
+    public interface ISavedFilterValidator
+    {
+        Task<bool> ExistsWithSameTitleForPersonAsync(string title, CancellationToken cancellationToken);
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/SavedFilterValidators/ISavedFilterValidator.cs:8]: ExistsWithSameTitleForCurrentPersonAsync

File: src/Equinor.Procosys.Preservation.Command/Validators/SavedFilterValidators/SavedFilterValidator.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            _context = context;
+            _personRepository = personRepository;
+            _currentUserProvider = currentUserProvider;
+        }
+
+        public async Task<bool> ExistsWithSameTitleForPersonAsync(string title, CancellationToken cancellationToken)
+        {
+            var currentUser = await _personRepository.GetByOidAsync(_currentUserProvider.GetCurrentUserOid());
+
+            return await (from p in _context.QuerySet<Person>()
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/SavedFilterValidators/SavedFilterValidator.cs:30]: Have you tested this in swagger? I'm not sure this work with p.SavedFilters.Any(sf => sf.Title == title)
Prefer to rewrite to use same pattern as when validating of a step with a title exists in a journey

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/PersonsController.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+{
+    [ApiController]
+    [Route("Persons")]
+    public class PersonsController : ControllerBase
+    {
+        private readonly IMediator _mediator;
+
+        public PersonsController(IMediator mediator) => _mediator = mediator;
+
+        [Authorize(Roles = Permissions.PRESERVATION_PLAN_CREATE)]
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/PersonsController.cs:21]: other than planners also need to save filters

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/CreateSavedFilterDto.cs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Equinor.Procosys.Preservation.WebApi.Controllers.Persons
+{
+    public class CreateSavedFilterDto
+    {
+        public string Title { get; set; }
+        public string Criteria { get; set; }
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/CreateSavedFilterDto.cs:6]: Has this been discussed in team? With Arild? Should client be allowed to store whatever string in Criteria?
This solution put all logic to client, which need to serialize and deserialize saved data in some known pattern.

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/PersonsController.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        [HttpPost("/SavedFilter")]
+        public async Task<ActionResult<int>> CreateSavedFilter(
+            [FromHeader(Name = CurrentPlantMiddleware.PlantHeader)]
+            [Required]
+            [StringLength(PlantEntityBase.PlantLengthMax, MinimumLength = PlantEntityBase.PlantLengthMin)]
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/PersonsController.cs:26]: remove

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/CreateSavedFilterDto.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Equinor.Procosys.Preservation.WebApi.Controllers.Persons
+{
+    public class CreateSavedFilterDto
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/CreateSavedFilterDto.cs:3]: Need validator class to check maxlength and required for both Title and Criteria

File: src/Equinor.Procosys.Preservation.Command/Validators/SavedFilterValidators/SavedFilterValidator.cs (hunk 4 above)
Comment (llykkelarsen) [reply src/Equinor.Procosys.Preservation.Command/Validators/SavedFilterValidators/SavedFilterValidator.cs:30]: Works fine but will rewrite for consistency.

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/CreateSavedFilterDto.cs (hunk 6 above)
Comment (llykkelarsen) [reply src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/CreateSavedFilterDto.cs:6]: Has been discussed with Arild and Pål. Did you have a different implementation in mind?

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/PersonsController.cs [hunk 9]
```diff
@@ -18,12 +18,11 @@ public class PersonsController : ControllerBase
 
         public PersonsController(IMediator mediator) => _mediator = mediator;
 
-        [Authorize(Roles = Permissions.PRESERVATION_PLAN_CREATE)]
+        [Authorize(Roles = Permissions.PRESERVATION_CREATE)]
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/PersonsController.cs:21]: Need any of roles PRESERVATION_PLAN_CREATE, PRESERVATION_CREATE
I think we have such endpoint somewhere else, Or'ing 2 roles

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/CreateSavedFilterDto.cs (hunk 6 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/CreateSavedFilterDto.cs:6]: Nahh .. I would have written strong typed classes which client need to fill exactly and send to endpoint. The backend handler then serialize data to JSON. We try your solution for now

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/PersonsController.cs (hunk 9 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.WebApi/Controllers/Persons/PersonsController.cs:21]: Don't bother with this. Planners should have both roles so I guess we will be covered

**Pull request merged**